plugins/outputs/opensearch/opensearch.go

```diff
@@ -0,0 +1,454 @@
+	"strconv"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/opensearch-project/opensearch-go/v2"
+	"github.com/opensearch-project/opensearch-go/v2/opensearchapi"
+	"github.com/opensearch-project/opensearch-go/v2/opensearchutil"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/internal/choice"
+	httpconfig "github.com/influxdata/telegraf/plugins/common/http"
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+type Opensearch struct {
+	Username            config.Secret `toml:"username"`
+	Password            config.Secret `toml:"password"`
+	AuthBearerToken     config.Secret `toml:"auth_bearer_token"`
+	EnableGzip          bool          `toml:"enable_gzip"`
+	EnableSniffer       bool          `toml:"enable_sniffer"`
+	FloatHandling       string        `toml:"float_handling"`
+	FloatReplacement    float64       `toml:"float_replacement_value"`
+	ForceDocumentID     bool          `toml:"force_document_id"`
+	IndexName           string        `toml:"index_name"`
+	TemplateName        string        `toml:"template_name"`
+	ManageTemplate      bool          `toml:"manage_template"`
+	OverwriteTemplate   bool          `toml:"overwrite_template"`
+	pipelineName        string
+	DefaultPipeline     string `toml:"default_pipeline"`
+	UsePipeline         string `toml:"use_pipeline"`
+	pipelineTagKeys     []string
+	tagKeys             []string
+	DefaultTagValue     string          `toml:"default_tag_value"`
+	Timeout             config.Duration `toml:"timeout"`
+	HealthCheckInterval config.Duration `toml:"health_check_interval"`
+	HealthCheckTimeout  config.Duration `toml:"health_check_timeout"`
+	URLs                []string        `toml:"urls"`
+	Log                 telegraf.Logger `toml:"-"`
+	onSucc              func(context.Context, opensearchutil.BulkIndexerItem, opensearchutil.BulkIndexerResponseItem)
+	onFail              func(context.Context, opensearchutil.BulkIndexerItem, opensearchutil.BulkIndexerResponseItem, error)
+	configOptions       httpconfig.HTTPClientConfig
+	osClient            *opensearch.Client
+}
+
+//go:embed template.json
+var indexTemplate string
+
+type templatePart struct {
+	TemplatePattern string
+}
+
+func (*Opensearch) SampleConfig() string {
+	return sampleConfig
+}
+
+func (o *Opensearch) Init() error {
+	if len(o.URLs) == 0 || o.IndexName == "" {
+		return fmt.Errorf("opensearch urls or index_name is not defined")
+	}
+
+	// Determine if we should process NaN and inf values
+	valOptions := []string{"", "none", "drop", "replace"}
+	if err := choice.Check(o.FloatHandling, valOptions); err != nil {
+		return fmt.Errorf("invalid float_handling type %q", o.FloatHandling)
+	}
+
+	if o.FloatHandling == "" {
+		o.FloatHandling = "none"
+	}
+
+	o.IndexName, o.tagKeys = o.GetReplacementKeys(o.IndexName, ".Tag", "%s")
+	o.pipelineName, o.pipelineTagKeys = o.GetReplacementKeys(o.UsePipeline, "", "%s")
+
```

Comment:
Please replace all your manual, hand-crafted templating using [Golang text/templates](https://pkg.go.dev/text/template#pkg-examples). Just compile the template here using
```golang
o.indexTmpl, err = template.New("index").Parse(o.IndexName)
```
and later use the resulting template in `Write` by calling `o.indexTmpl.Execute(&buf, m)`. Check-out the [lookup processor](https://github.com/influxdata/telegraf/blob/master/plugins/processors/lookup/lookup.go) on how to use metrics with the template...

Reply:
Thanks for the reference @srebhan, will try to make the changes in coming week.

Reply:
Hi @srebhan, made the changes, apologies for the delay again, please verify )